# Microbial features filtering
This tutorial is to use a python script to pre-filter metagenome-profiled features (i.e., taxa, genes and metabolic pathways), controlling relative abundances and prevalence.

#### Python packages required

* [pandas >= 1.3.5](https://pandas.pydata.org/)
* [numpy >= 1.20.3](https://numpy.org/)

#### Microbial features filtering using `profile_cleaner.py`

Here, we introduce a python script `profile_cleaner.py` from `path_to_the_package/KunDH-2024-CR-DietMetagenomics/scripts` to filter out low abundance and low prevalence features (e.g, taxa, genes and metabolic pathways):

```{python}
usage: profile_cleaner.py [-h] [--profile [PROFILE]] [--row_numbers [ROW_NUMBERS]] [--min_abundance [MIN_ABUNDANCE]] [--min_prevalence [MIN_PREVALENCE]]
                          [--nth_percentile [NTH_PERCENTILE]] [--output [OUTPUT]]

This program is to clean microbiome profiles.

optional arguments:
  -h, --help            show this help message and exit
  --profile [PROFILE]   Input the merged abundance profile generated by MetaPhlAn or HUMAnNN.
  --row_numbers [ROW_NUMBERS]
                        If you have metadata wedged please specify row numbers delimited by comma, zero-based and without considering header, e.g 0,1,2,3,4.
                        Otherwise leave this as it is.
  --min_abundance [MIN_ABUNDANCE]
                        Specify the minimum abundance. Default: [0]
  --min_prevalence [MIN_PREVALENCE]
                        Specify the minimum prevalence in all samples (in percentage). Default: [0]. Note: if this parameter is set --nth_percentitle will be
                        muted.
  --nth_percentile [NTH_PERCENTILE]
                        Specify the minimum relative abundance at the nth percentitle. Default: [10]. Note: if this parameter is set --min_prevalence will be
                        muted.
  --output [OUTPUT]     Give a name to the output file.

examples: profile_cleaner.py --profile <abundance_table.tsv> --row_numbers <0,1,2> --min_abundance <minimum_abundance> --min_prevalence <minimum_prevalence> --output <abundance_table_filtered.tsv>
```

#### Demo with example data

The example input, a [metaphlan taxonomy table](../example_data/microbial_species_prefiltering.tsv), can be found in `path_to_the_package/KunDH-2024-CR-DietMetagenomics/example_data`.  


Example command line:

~~~bash
profile_cleaner.py --profile microbial_species_prefiltering.tsv --min_abundance 0.01 --min_prevalence 10 --output microbial_species_postfiltering.tsv
~~~

The output, [filtered taxonomy table](../example_data/microbial_species_postfiltering.tsv), can be found in `path_to_the_package/KunDH-2024-CR-DietMetagenomics/example_data`. 